File: src/main/java/com/github/fligneul/debtplugin/debt/glutter/DebtGutterLineMarkerProvider.java
package com.github.fligneul.debtplugin.debt.glutter;

import com.github.fligneul.debtplugin.debt.icons.DebtIcons;
import com.github.fligneul.debtplugin.debt.model.DebtItem;
import com.github.fligneul.debtplugin.debt.model.Repository;
import com.github.fligneul.debtplugin.debt.service.DebtService;
import com.intellij.codeInsight.daemon.LineMarkerInfo;
import com.intellij.codeInsight.daemon.LineMarkerProvider;
import com.intellij.openapi.editor.Document;
import com.intellij.openapi.editor.markup.GutterIconRenderer;
import com.intellij.openapi.project.DumbAware;
import com.intellij.openapi.project.Project;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.intellij.psi.PsiWhiteSpace;
import com.intellij.psi.util.PsiTreeUtil;
import com.intellij.util.Function;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.Icon;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Optional;

public class DebtGutterLineMarkerProvider implements LineMarkerProvider, DumbAware {
    @Override
    public @Nullable LineMarkerInfo<?> getLineMarkerInfo(PsiElement element) {
        PsiFile file = element.getContainingFile();
        if (file == null) return null;
        Document doc = file.getViewProvider().getDocument();
        if (doc == null) return null;
        if (element.getTextRange() == null) return null;

        int lineNumber = doc.getLineNumber(element.getTextRange().getStartOffset());
        int lineStartOffset = doc.getLineStartOffset(lineNumber);

        PsiElement firstAtLine = file.findElementAt(lineStartOffset);
        if (firstAtLine == null) return null;
        PsiElement firstNonWs = nextNonWhitespace(firstAtLine);
        if (firstNonWs == null) return null;

        if (element != firstNonWs) return null;

        Project project = element.getProject();
        DebtService debtService = project.getService(DebtService.class);

        if (file.getVirtualFile() == null) return null;
        String osPath = file.getVirtualFile().getPath();
        String basePath = debtService.findRepoRootForAbsolutePath(osPath);
        final Map.Entry<Repository, List<DebtItem>> debtsForRepository = debtService.getDebtForRepositoryAbsolutePath(basePath).orElseThrow();

        String currentRel = toProjectRelative(osPath, basePath);
        int lineInFile = lineNumber + 1;
        List<DebtItem> debtsOnLine = new ArrayList<>();
        for (DebtItem debtItem : debtsForRepository.getValue()) {
            String stored = debtItem.getFile();
            if (stored.isBlank()) continue;
            String storedRel = toProjectRelative(stored, basePath);
            if (storedRel.equalsIgnoreCase(currentRel) && debtItem.getLine() == lineInFile) {
                debtsOnLine.add(debtItem);
            }
        }
        if (debtsOnLine.isEmpty()) return null;

        Icon icon = DebtIcons.TECHNICAL_DEBT;
        Function<PsiElement, String> tooltipProvider = psi -> {
            if (debtsOnLine.size() == 1) {
                DebtItem d = debtsOnLine.get(0);
                return """
                            Debt
                                %s
                                by %s
                        """.formatted(getDebtName(d), d.getUsername());
            } else {
                StringBuilder sb = new StringBuilder();
                for (DebtItem it : debtsOnLine) {
                    sb.append("Debt : \n ")
                            .append(getDebtName(it))
                            .append(" \n by ")
                            .append(it.getUsername())
                            .append("\n");
                }
                return sb.toString();
            }
        };

        return new LineMarkerInfo<>(
                element,
                firstNonWs.getTextRange(),
                icon,
                tooltipProvider,
                (evt, el) -> {
                },
                GutterIconRenderer.Alignment.LEFT,
                () -> "Debt"
        );
    }

    private @NotNull String getDebtName(final DebtItem d) {
        final String debtName = Optional.ofNullable(d.getTitle())
                .filter(str -> !str.isBlank())
                .or(() -> Optional.ofNullable(d.getDescription())
                        .filter(str -> !str.isBlank()))
                .orElse("UNKNOWN");
        return debtName;
    }

    private static @Nullable PsiElement nextNonWhitespace(@Nullable PsiElement start) {
        PsiElement el = start;
        while (el instanceof PsiWhiteSpace) {
            el = PsiTreeUtil.nextLeaf(el, true);
        }
        if (el instanceof PsiWhiteSpace) return null;
        return el;
    }

    public void collectSlowLineMarkers(List<? extends PsiElement> elements, Collection<? super LineMarkerInfo<?>> result) {
        // Not used
    }

    private static String normalizePath(String path) {
        try {
            return Paths.get(path).toAbsolutePath().normalize().toString().replace('\\', '/').toLowerCase(Locale.getDefault());
        } catch (Exception e) {
            return path.replace('\\', '/').toLowerCase(Locale.getDefault());
        }
    }

    private static String toProjectRelative(String anyPath, String basePath) {
        if (anyPath == null) return "";
        String norm = anyPath.replace('\\', '/');
        try {
            Path p = Paths.get(anyPath);
            if (basePath != null) {
                Path base = Paths.get(basePath).toAbsolutePath().normalize();
                Path abs = p.isAbsolute() ? p.toAbsolutePath().normalize() : base.resolve(p).normalize();
                if (abs.startsWith(base)) {
                    return base.relativize(abs).toString().replace('\\', '/');
                }
                return abs.toString().replace('\\', '/');
            } else {
                Path abs = p.isAbsolute() ? p.toAbsolutePath().normalize() : p.normalize();
                return abs.toString().replace('\\', '/');
            }
        } catch (Exception e) {
            return norm;
        }
    }
}

File: src/main/kotlin/com/github/fligneul/debtplugin/debt/DebtGutterLineMarkerProvider.kt
package com.github.fligneul.debtplugin.debt

import com.intellij.codeInsight.daemon.LineMarkerInfo
import com.intellij.codeInsight.daemon.LineMarkerProvider
import com.intellij.openapi.components.service
import com.intellij.openapi.editor.markup.GutterIconRenderer
import com.intellij.openapi.project.Project
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiWhiteSpace
import com.intellij.psi.util.PsiTreeUtil
import com.intellij.util.Function
import javax.swing.Icon
import java.nio.file.Paths
import java.util.Locale
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.util.IconLoader

class DebtGutterLineMarkerProvider : LineMarkerProvider, com.intellij.openapi.project.DumbAware {
    private val log = Logger.getInstance(DebtGutterLineMarkerProvider::class.java)

    companion object {
        val DEBT_ICON: Icon = IconLoader.getIcon("/icons/technicalDebt.svg", DebtGutterLineMarkerProvider::class.java)
    }

    override fun getLineMarkerInfo(element: PsiElement): LineMarkerInfo<*>? {
        // Use slow markers flow for reliable per-line detection across languages
        return null
    }

    private fun nextNonWhitespace(start: PsiElement?): PsiElement? {
        var el = start
        // Move forward until a non-whitespace leaf/element is found
        while (el is PsiWhiteSpace) {
            el = PsiTreeUtil.nextLeaf(el, true)
        }
        // Some languages may return composite elements at line start; skip pure whitespace children
        if (el is PsiWhiteSpace) return null
        return el
    }

    override fun collectSlowLineMarkers(elements: MutableList<out PsiElement>, result: MutableCollection<in LineMarkerInfo<*>>) {
        if (elements.isEmpty()) return
        val file = elements.first().containingFile ?: return
        val doc = file.viewProvider.document ?: return
        val project: Project = elements.first().project
        val debtService = project.service<DebtService>()

        val vFile = file.virtualFile ?: return
        val normFilePath = normalizePath(vFile.path)

        // Group elements by line and pick a representative non-whitespace element per line
        val firstElementPerLine = HashMap<Int, PsiElement>()
        for (el in elements) {
            val range = el.textRange ?: continue
            val line = doc.getLineNumber(range.startOffset)
            // pick the earliest element at this line
            val existing = firstElementPerLine[line]
            if (existing == null || range.startOffset < (existing.textRange?.startOffset ?: Int.MAX_VALUE)) {
                firstElementPerLine[line] = el
            }
        }

        for ((line, candidate) in firstElementPerLine) {
            val lineStart = doc.getLineStartOffset(line)
            val firstAtLine = file.findElementAt(lineStart) ?: candidate
            val anchor = nextNonWhitespace(firstAtLine) ?: candidate

            val lineInFile = line + 1
            val debtsOnLine = debtService.all().filter {
                normalizePath(it.file) == normFilePath && it.line == lineInFile
            }
            if (debtsOnLine.isEmpty()) continue

            val icon: Icon = DEBT_ICON
            val tooltipProvider = Function<PsiElement, String> {
                if (debtsOnLine.size == 1) {
                    val d = debtsOnLine.first()
                    "Debt: ${'$'}{d.description} [${'$'}{d.status}, ${'$'}{d.priority}]"
                } else {
                    val items = debtsOnLine.joinToString("\n") { "- ${'$'}{it.description} [${'$'}{it.status}, ${'$'}{it.priority}]" }
                    "Debts (${debtsOnLine.size}):\n${items}"
                }
            }

            result.add(
                LineMarkerInfo(
                    anchor,
                    anchor.textRange,
                    icon,
                    tooltipProvider,
                    { _, _ -> },
                    GutterIconRenderer.Alignment.LEFT,
                    { "Debt" }
                )
            )
        }
    }

    private fun normalizePath(path: String): String {
        return try {
            Paths.get(path).toAbsolutePath().normalize().toString().replace('\\', '/').lowercase(Locale.getDefault())
        } catch (e: Exception) {
            path.replace('\\', '/').lowercase(Locale.getDefault())
        }
    }
}
